src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -0,0 +1,1144 @@
+            {
+                throw new HttpRequestException("Invalid characters in response line");
+            }
+
+            _sb.Clear();
+
+            // Parse reason phrase
+            char c = await ReadCharAsync(cancellationToken);
+            while (c != '\r')
+            {
+                _sb.Append(c);
+                c = await ReadCharAsync(cancellationToken);
+            }
+
+            if (await ReadCharAsync(cancellationToken) != '\n')
+            {
+                throw new HttpRequestException("Saw CR without LF while parsing response line");
+            }
+
+            response.ReasonPhrase = _sb.ToString();
+
+            var responseContent = new HttpConnectionContent(CancellationToken.None);
+
+            // Parse headers
+            _sb.Clear();
+            c = await ReadCharAsync(cancellationToken);
+            while (true)
+            {
+                if (c == '\r')
+                {
+                    if (await ReadCharAsync(cancellationToken) != '\n')
+                    {
+                        throw new HttpRequestException("Saw CR without LF while parsing headers");
+                    }
+
+                    break;
+                }
+
+                // Get header name
+                while (c != ':')
+                {
+                    _sb.Append(c);
+                    c = await ReadCharAsync(cancellationToken);
+                }
+
+                string headerName = _sb.ToString();
+
+                _sb.Clear();
+
+                // Get header value
+                c = await ReadCharAsync(cancellationToken);
+                while (c == ' ')
+                {
+                    c = await ReadCharAsync(cancellationToken);
+                }
+
+                while (c != '\r')
+                {
+                    _sb.Append(c);
+                    c = await ReadCharAsync(cancellationToken);
+                }
+
+                if (await ReadCharAsync(cancellationToken) != '\n')
+                {
+                    throw new HttpRequestException("Saw CR without LF while parsing headers");
+                }
+
+                string headerValue = _sb.ToString();
+
+                // TryAddWithoutValidation will fail if the header name has trailing whitespace.
+                // So, trim it here.
+                // TODO: Not clear to me from the RFC that this is really correct; RFC seems to indicate this should be an error.
+                // However, tests claim this is important for compat in practice.
+                headerName = headerName.TrimEnd();
+
+                // Add header to appropriate collection
+                // Don't ask me why this is the right API to call, but apparently it is
+                if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                {
+                    if (!responseContent.Headers.TryAddWithoutValidation(headerName, headerValue))
```

Comment:
This distinction is endemic to HttpClient, for better or worse.  But yeah, the comment here should be better.